zipkin-storage/cassandra3/src/main/java/zipkin/storage/cassandra3/CassandraSpanStore.java

```diff
@@ -468,21 +456,101 @@
               .setUUID("end_ts", UUIDs.endOf(endTsMillis))
               .setInt("limit_", limit);
 
-      return transform(session.executeAsync(bound),
-          new Function<ResultSet, Map<TraceIdUDT, Long>>() {
-            @Override public Map<TraceIdUDT, Long> apply(ResultSet input) {
-              Map<TraceIdUDT, Long> traceIdsToTimestamps = new LinkedHashMap<>();
-              for (Row row : input) {
-                traceIdsToTimestamps.put(
-                        row.get("trace_id", TraceIdUDT.class),
-                        row.getLong("ts"));
-              }
-              return traceIdsToTimestamps;
-            }
-          }
-      );
+      return transform(session.executeAsync(bound), readResultsAsMap(Collections.emptyMap(), traceIdToLong));
     } catch (RuntimeException ex) {
       return immediateFailedFuture(ex);
     }
   }
+
+  private static <K, T> AsyncFunction<ResultSet, Map<K, T>> readResultsAsMap(final Map<K, T> results,
+      Function<Row, Map.Entry<K, T>> rowMapper) {
+    return new AsyncFunction<ResultSet, Map<K, T>>() {
+      @Override public ListenableFuture<Map<K, T>> apply(ResultSet rs) throws Exception {
+        // How far we can go without triggering the blocking fetch:
+        int remainingInPage = rs.getAvailableWithoutFetching();
```

Comment:
I'm a bit late to the party on this one.

Would it not be simpler to call `rs.fetchMoreResults()` ?

`fetchMoreResults()` is an async call, and improves the chances that the next page is ready to go by the time this page is exhausted.

Taking this approach is simpler as it removes the concurrency/futures from this code, allowing it to be just the single iteration loop